### PR TITLE
Add fused MoE support to SelectiveMixedPrecision heuristics

### PR DESCRIPTION
**Pull request by titaiwangms**
## Describe your changes

Add explicit, fail-closed fused-MoE support to the fixed `SelectiveMixedPrecision` heuristics.

- Add `moe: true` support to `high_precision_mlp_down` and `high_precision_mlp_down_qkv` for explicitly recognized Qwen3/Qwen3.5 fused K-last expert layouts.
- Resolve the routed expert output projection through `LayerWrapper` and canonical `iter_quant_targets` identities, emitting whole per-layer `experts.down_proj` overrides rather than per-expert precision.
- Keep `high_precision_lm_head` as a legal MoE no-op, and reject score-based algorithms with `moe: true` until their parameter-level scoring is implemented separately.
- Add `mixed_precision_info.requires_moe` and require an explicit `moe: true` on the first MoE-capable RTN, GPTQ, or KQuant consumer. Validate that required expert overrides survive skip/component selection and are actually selected or already materialized before mutating parameters.
- Preserve AutoClip and follow-up category-pass composition, prevent persisted metadata from enabling/disabling the consumer’s MoE setting, and reject malformed or stale MoE metadata.
- Keep routers and shared-expert gates unquantized; shared-expert MLP weights remain at the default precision. ModuleList experts, unknown/transposed layouts, score-based MoE selection, and ONNX/Mobius export are intentionally out of scope.

This is the first implementation stage for #2638. A follow-up will generalize SNR/IQE scoring to fused parameter targets; KLD-gradient requires separate routing/gradient/memory design.

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`.
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

Release note: `SelectiveMixedPrecision` fixed heuristics can now plan higher precision for supported fused Qwen MoE expert output projections with explicit downstream MoE opt-in and fail-closed validation.

## (Optional) Issue link

Part of #2638.

### CODE REVIEW

File: olive/passes/pytorch/quant_utils.py [hunk 1]
```diff
@@ -495,6 +498,177 @@ def _collect_already_quantized_names(model: torch.nn.Module) -> set[str]:
...
+        for module, pname, full_name in iter_quant_targets(
+            root_model,
+            quantize_lm_head=True,
+            quantize_embeds=True,
+            quantize_moe=True,
+            quantize_vision=True,
+            skip_already_quantized=False,
+        )
+        if not isinstance(module, (torch.nn.Linear, torch.nn.Embedding)) and module._parameters[pname].dim() == 3
+    }
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply olive/passes/pytorch/quant_utils.py]: Updated in e4176f98. The target scan now uses `_parameters.get(pname)` and only inspects the dimension after confirming the parameter is non-None.

File: olive/passes/pytorch/selective_mixed_precision.py [hunk 2]
```diff
@@ -876,22 +929,146 @@ def get_high_precision_config(
...
+                fused_expert_override_emitted = True
+            else:
+                for output_name in dense_output_names[layer_idx]:
+                    overrides[f"{layer_prefix}.{layer_idx}.{output_name}"] = override_config
+
+        return {"bits": bits}, overrides, fused_expert_override_emitted
+
+    @staticmethod
+    def _is_selected_heuristic_layer(layer_idx: int, num_layers: int) -> bool:
+        return layer_idx < num_layers / 8 or layer_idx >= 7 * num_layers / 8 or ((layer_idx - num_layers // 8) % 3 == 2)
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply olive/passes/pytorch/selective_mixed_precision.py]: Updated in e4176f98. Wrapped the layer-selection predicate across multiple lines without changing its semantics.

File: olive/passes/pytorch/quant_utils.py [hunk 3]
```diff
@@ -495,6 +498,182 @@ def _collect_already_quantized_names(model: torch.nn.Module) -> set[str]:
...
+    ):
+        parameter = module._parameters.get(pname)
+        if (
+            parameter is not None
+            and not isinstance(module, (torch.nn.Linear, torch.nn.Embedding))
+            and parameter.dim() == 3
+        ):
+            fused_expert_targets.add(full_name)
+
+    return {name for name in mp_info["overrides"] if name in fused_expert_targets}
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply olive/passes/pytorch/quant_utils.py]: Fixed in 6d88cb5d. Required-MoE plans now validate every override against the complete canonical quantization target set before extracting fused expert targets, so stale names cannot be silently dropped. The same change also verifies immutable QuantTensor precision against both the effective Olive config and the required override, and defers embedding untying until validation succeeds.

File: docs/source/features/quantization.md [hunk 4]
```diff
@@ -226,6 +226,36 @@ are quantized; expert biases remain in full precision.
...
+
+`requires_moe` is pass-emitted metadata, not an additional return value from the
+lower-level `get_high_precision_config` or deprecated `get_k_quant_config`
+helpers; both retain their two-value return contract.
+
+This planning metadata belongs to the Hugging Face/PyTorch pass boundary.
+Plain ONNX quantization does not consume it. For Mobius / ORT GenAI
+`ModelBuilder`, first materialize the plan with a supported PyTorch quantizer so
+the resulting Olive Hugging Face `quantization_config` records the selected
+expert precisions; planner metadata alone is not a Mobius input contract.
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply docs/source/features/quantization.md]: Fixed in 999c47c6. The documentation now explicitly limits this stage to materializing the plan into an Olive Hugging Face/PyTorch checkpoint and states that Mobius / ORT GenAI ModelBuilder export is out of scope and has not been validated.

**Review by jambayk — APPROVED** [review]
